### PR TITLE
feat: add support for `std::variant`

### PR DESCRIPTION
**Pull request by uyha**

### CODE REVIEW

**Comment by uyha** [comment]
@redboltz I'm having a bit of problem implementing `object` and `object_with_zone`, I'm not actually sure what they do and what operation is safe to do in the `object` case.

**Comment by redboltz** [comment]
> @redboltz I'm having a bit of problem implementing `object` and `object_with_zone`, I'm not actually sure what they do and what operation is safe to do in the `object` case.

`object` adaptor is only provided the packed type can be expressed by `msgpack::object` without zone allocation.

In the case of `std::variant` T is unpredictable. It is the same as `std::optional`.

https://github.com/msgpack/msgpack-c/blob/ac062e28cbf80db345800b6e113dea8ab0055499/include/msgpack/v1/adaptor/cpp17/optional.hpp#L67-L81

`std::optional` adaptor provides `object` adaptor. If `T` has object adator, then forward it, otherwise compile error.
At first, I thought that `std::variant` is the same case, but I noticed that this adaptor needs index. So MessagePack ARRAY is required. That means zone allocation is required.

Hence, this `std::variant` adaptor shouldn't provide `object` adaptor.

**Comment by redboltz** [comment]
There are some of coding style fix request.
(Sorry for lack of documents)

I will add the comment inline later. Maybe we need to communicate several times.

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#include "msgpack/adaptor/adaptor_base.hpp"
+#include "msgpack/object.hpp"
+#include "msgpack/versioning.hpp"
+
+#include <variant>
+
+namespace msgpack {
+MSGPACK_API_VERSION_NAMESPACE(v1) {
+  namespace adaptor {
+  namespace detail {
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
namespace msgpack {
MSGPACK_API_VERSION_NAMESPACE(v1) {
namespace adaptor {
namespace detail {
```

namespace doesn't introduce indent.

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  template <typename Variant, typename T, typename... Ts, std::size_t current_index, std::size_t... indices>
+  Variant construct_variant(std::size_t index,
+                            msgpack::object &object,
+                            std::index_sequence<current_index, indices...>) {
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
Variant construct_variant(
    std::size_t index,
    msgpack::object& object,
    std::index_sequence<current_index, indices...>) {
```

When the parameter list is multiple lines, please do like this. If the function name is long, this approach can make much areas.

Please connect `&` `*` to the type.
I know `int& i, j;` is misleading. But in this case we use

```cpp
int& i;
int& j;
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    if constexpr(sizeof...(Ts) == 0) {
+      return object.as<T>();
+    } else {
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
    } 
    else {
```

Please use this style. Not 100% complied, but at least new code do that.

**Review by redboltz — CHANGES_REQUESTED** [review]
Please use 4 whitespaces indent.

I pointed out one case for each request. Please fix other same parts.

**Comment by uyha** [comment]
@redboltz could you please have a look again?

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+namespace msgpack {
+MSGPACK_API_VERSION_NAMESPACE(v1) {
+namespace adaptor {
+namespace detail {
+template <
+    typename Variant,
+    typename T,
+    typename... Ts,
+    std::size_t current_index,
+    std::size_t... indices>
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
    std::size_t... indices
>
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        obj = msgpack::object(value, zone);
+    }
+
+    msgpack::object& obj;
+    msgpack::zone& zone;
+};
+} // namespace detail
+
+template <typename... Ts>
+struct as<std::variant<Ts...>, typename std::enable_if<(msgpack::has_as<Ts>::value && ...)>::type> {
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp:73]:
```suggestion
struct as<std::variant<Ts...>, typename std::enable_if<(msgpack::has_as<Ts>::value&& ...)>::type> {
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    std::variant<Ts...> operator()(msgpack::object const& o) const {
+        if (  o.type != msgpack::type::ARRAY
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp:75]:
```suggestion
        if (   o.type != msgpack::type::ARRAY
```

It is not a part of coding style but I guess you want to do like this.

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        return object.as<T>();
+    }
+    else {
+        if (index == current_index) {
+            return object.as<T>();
+        }
+        return construct_variant<Variant, Ts...>(
+                   index,
+                   object,
+                   std::index_sequence<indices...>());
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
        return construct_variant<Variant, Ts...>(
            index,
            object,
            std::index_sequence<indices...>()
        );
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            || o.via.array.ptr[0].type != msgpack::type::POSITIVE_INTEGER
+            || o.via.array.ptr[0].via.u64 >= sizeof...(Ts)) {
+            throw msgpack::type_error{};
+        }
+
+        return detail::construct_variant<std::variant<Ts...>, Ts...>(
+          o.via.array.ptr[0].as<std::size_t>(),
+          o.via.array.ptr[1],
+          std::make_index_sequence<sizeof...(Ts)>()
+        );
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
        return detail::construct_variant<std::variant<Ts...>, Ts...>(
            o.via.array.ptr[0].as<std::size_t>(),
            o.via.array.ptr[1],
            std::make_index_sequence<sizeof...(Ts)>()
        );
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    }
+};
+
+template<typename... Ts>
+struct convert<std::variant<Ts...>> {
+    msgpack::object const& operator()(msgpack::object const& o, std::variant<Ts...>& v) const {
+        if (  o.type != msgpack::type::ARRAY
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp:93]:
```suggestion
        if (   o.type != msgpack::type::ARRAY
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            || o.via.array.ptr[0].type != msgpack::type::POSITIVE_INTEGER
+            || o.via.array.ptr[0].via.u64 >= sizeof...(Ts)) {
+            throw msgpack::type_error{};
+        }
+
+        v =  detail::construct_variant<std::variant<Ts...>, Ts...>(
+          o.via.array.ptr[0].as<std::size_t>(),
+          o.via.array.ptr[1],
+          std::make_index_sequence<sizeof...(Ts)>()
+        );
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
        v =  detail::construct_variant<std::variant<Ts...>, Ts...>(
            o.via.array.ptr[0].as<std::size_t>(),
            o.via.array.ptr[1],
            std::make_index_sequence<sizeof...(Ts)>()
        );
```

File: test/msgpack_cpp17.cpp [hunk 11]
```diff
@@ -461,4 +461,29 @@ BOOST_AUTO_TEST_CASE(carray_byte_object_with_zone)
...
+BOOST_AUTO_TEST_CASE(variant_pack_unpack_as) {
+  std::stringstream ss;
+  std::variant<bool, int, float, double> val1{1.0};
+  msgpack::pack(ss, val1);
+  std::string const& str = ss.str();
+  msgpack::object_handle oh =
+      msgpack::unpack(str.data(), str.size());
+  std::variant<bool, int, float, double> val2 = oh.get().as<std::variant<bool, int, float, double> >();
+  BOOST_CHECK(val1 == val2);
+  BOOST_CHECK_THROW((oh.get().as<std::variant<bool>>()), msgpack::type_error);
```
Comment (redboltz) [inline test/msgpack_cpp17.cpp]:
```suggestion
    std::stringstream ss;
    std::variant<bool, int, float, double> val1{1.0};
    msgpack::pack(ss, val1);
    std::string const& str = ss.str();
    msgpack::object_handle oh =
        msgpack::unpack(str.data(), str.size());
    std::variant<bool, int, float, double> val2 = oh.get().as<std::variant<bool, int, float, double> >();
    BOOST_CHECK(val1 == val2);
    BOOST_CHECK_THROW((oh.get().as<std::variant<bool>>()), msgpack::type_error);
```

File: test/msgpack_cpp17.cpp [hunk 12]
```diff
@@ -461,4 +461,29 @@ BOOST_AUTO_TEST_CASE(carray_byte_object_with_zone)
...
+}
+
+BOOST_AUTO_TEST_CASE(variant_with_zone) {
+  msgpack::zone z;
+  std::variant<bool, int, float, double> val1{1.0};
+  msgpack::object obj(val1, z);
+  std::variant<bool, int, float, double> val2 = obj.as<std::variant<bool, int, float, double>>();
+  BOOST_CHECK(val1 == val2);
+  BOOST_CHECK_THROW((obj.as<std::variant<bool>>()), msgpack::type_error);
```
Comment (redboltz) [inline test/msgpack_cpp17.cpp]:
```suggestion
    msgpack::zone z;
    std::variant<bool, int, float, double> val1{1.0};
    msgpack::object obj(val1, z);
    std::variant<bool, int, float, double> val2 = obj.as<std::variant<bool, int, float, double>>();
    BOOST_CHECK(val1 == val2);
    BOOST_CHECK_THROW((obj.as<std::variant<bool>>()), msgpack::type_error);
```

File: include/msgpack/v1/adaptor/cpp17/variant.hpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    }
+};
+
+
+template<typename... Ts>
+struct object_with_zone<std::variant<Ts...>> {
+    void operator()(msgpack::object::with_zone& o, std::variant<Ts...> const& v) const {
+        msgpack::object *p = static_cast<msgpack::object *>(
+                                 o.zone.allocate_align(sizeof(msgpack::object) * 2,
+                                 MSGPACK_ZONE_ALIGNOF(msgpack::object)));
```
Comment (redboltz) [inline include/msgpack/v1/adaptor/cpp17/variant.hpp]:
```suggestion
msgpack::object* p =
    static_cast<msgpack::object*>(
        o.zone.allocate_align(sizeof(msgpack::object) * 2,
        MSGPACK_ZONE_ALIGNOF(msgpack::object))
    );
```

**Review by redboltz — CHANGES_REQUESTED** [review]
I've checked updated one.

File: include/msgpack/v1/adaptor/cpp17/variant.hpp (hunk 6 above)
Comment (uyha) [reply include/msgpack/v1/adaptor/cpp17/variant.hpp:75]: yes, I want the coditions to be aligned

File: include/msgpack/v1/adaptor/cpp17/variant.hpp (hunk 5 above)
Comment (uyha) [reply include/msgpack/v1/adaptor/cpp17/variant.hpp:73]: but this is the boolean and operator, not an rvalue reference.

Comment (redboltz) [reply include/msgpack/v1/adaptor/cpp17/variant.hpp:73]: > but this is the boolean and operator, not an rvalue reference.

Aha, sorry I misunderstood about that. You original code is right. Please revert it.

**Comment by uyha** [comment]
@redboltz are this PR and #1076 good enough for merge?

**Review by redboltz — APPROVED** [review]

**Comment by redboltz** [comment]
Merged. Thank you!

**Pull request merged**